compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs

```diff
@@ -402,3 +380,141 @@ impl<'tcx> ConstToPat<'tcx> {
         Ok(Box::new(Pat { span, ty, kind }))
     }
 }
+
+/// Given a type with type parameters, visit every ADT looking for types that need to
+/// `#[derive(PartialEq)]` to be a structural type.
+fn extend_type_not_partial_eq<'tcx>(
```

Comment:
- This could also be used for the non-fallthrough case (where the top-level type isn't structural), as the visitor would work for that too.
- We could rework this to not operate on the `Diag` directly, instead create a new `Subdiagnostic` type that gets added to the `Diagnostic` as a `Vec`.